### PR TITLE
CORGI-857 cpe per component

### PR DESCRIPTION
**Pull request by jasinner**
I did this hoping that it would help with CORGI-602 but unfortunately I didn't help. However this does extend the work we did in CORGI-872 to work with streams which don't use errata_info to define which builds make them up.

### CODE REVIEW

**Comment by jasinner** [comment]
@RedHatProductSecurity/corgi-devs please review

File: corgi/core/models.py [hunk 1]
```diff
@@ -1774,25 +1775,30 @@ def get_roots(self, using: str = "read_only") -> QuerySet[ComponentNode]:
...
+                self.productvariants.exclude(cpe="")
+                .values_list("cpe", flat=True)
+                .distinct()
+                .order_by("cpe")
+            )
+        elif self.software_build:
+            build_variants = (
+                ProductComponentRelation.objects.filter(software_build=self.software_build)
+                .values_list("product_ref", flat=True)
+                .distinct()
```
Comment (juspence) [inline corgi/core/models.py]: This is finding all relations, but I think we should filter only for relations that refer to variants:
```suggestion
                ProductComponentRelation.objects.filter(software_build=self.software_build, type__in=ProductComponentRelation.VARIANT_TYPES)
                .values_list("product_ref", flat=True)
                .distinct()
```
This also requires that the relation be processed. That's fine if it's what you want. But we could also report CPEs even for Variant product_refs from unprocessed relations. We'd just need to filter the relations on `build_id=self.software_build.build_id, build_type=self.software_build.build_type` instead of using `software_build=self.software_build` directly.

File: corgi/core/models.py [hunk 2]
```diff
@@ -1774,25 +1775,30 @@ def get_roots(self, using: str = "read_only") -> QuerySet[ComponentNode]:
...
+            )
+            if build_variants:
+                return (
+                    CollectorErrataProductVariant.objects.filter(name__in=build_variants)
+                    .values_list("cpe", flat=True)
+                    .order_by("cpe")
```
Comment (juspence) [inline corgi/core/models.py:1803]: I thought we didn't want to access Collector models from Core model code. Is there a reason we don't just filter using the regular ProductVariant model? That should still return more data than self.productvariants. If we haven't saved the taxonomy for this component yet, self.productvariants would either be empty or incomplete / there will be no links to the models, but the ProductVariant models themselves will still exist and have all the right CPEs.

Whatever we do, we should also exclude empty "" string CPEs from being reported here.

File: corgi/core/models.py [hunk 3]
```diff
@@ -1774,25 +1775,30 @@ def get_roots(self, using: str = "read_only") -> QuerySet[ComponentNode]:
 
     @property
     def cpes(self) -> QuerySet:
-        """Build and return a list of CPEs from all Variants this Component relates to"""
```
Comment (juspence) [inline corgi/core/models.py:1780]: You deleted "from all Variants" in the new code. But aren't we still getting CPE values only from Variants? Or did you plan to eventually do something like below, in addition to variants?
1. From the relations table, look up all product_refs for this component
2. Filter only the stream-type relations and look up the related streams
3. Report CPEs for all the stream's child Variants, using each stream's "cpes" property

File: corgi/web/templates/product_manifest.json [hunk 4]
```diff
@@ -3,12 +3,16 @@
     {
       "copyrightText": {% if component.copyright_text %}"{{component.copyright_text|escapejs}}"{% else %}"NOASSERTION"{% endif %},
       "downloadLocation": {% if component.download_url %}"{{component.download_url|escapejs}}"{% else %}"NOASSERTION"{% endif %},
+
       "externalRefs": [{% for cpe in component.cpes %}
+        {# Make sure any component cpe is found in the product cpe list #}
+        {% if cpe in cpes %}
```
Comment (juspence) [inline corgi/web/templates/product_manifest.json]: Where is this CPEs variable defined? Also, I think you need to update the component_manifest.json template as well, right?

Finally, the for loop, comment, and if statement need to all be on the same line. This also applies to the endif / endfor bits below. Otherwise whitespace will end up in the template and it may break the JSON formatting.

**Review by juspence — COMMENTED** [review]
I'm not sure I understand the change, it seems incomplete. Right now we only report CPEs for root components, but I thought in the ticket we wanted to report CPEs for all components.

If that's correct, you can just add the `{% for cpe in component.cpes %}` logic to the externalRefs key for provided and upstream components as well. Right now this for-loop only exists for root components. This should be enough to start reporting CPEs for all components, changes to the Component model's "cpes" property shouldn't be needed.

Your changes to the "cpes" property will mostly work fine I think, but I'm not sure what edge-case you're trying to handle. I say "mostly" because self.software_build is only set for root components, and again I thought we were trying to report CPEs for all components.

File: tests/test_manifests.py [hunk 5]
```diff
@@ -132,8 +134,12 @@ def test_component_manifest_backslash():
 
 
 @pytest.mark.django_db(databases=("default", "read_only"), transaction=True)
-def a_test_slim_rpm_in_containers_manifest(stored_proc):
-    containers, stream, rpm_in_container = setup_products_and_rpm_in_containers()
```
Comment (juspence) [inline tests/test_manifests.py:136]: Do you know when we changed this test to be prefixed with `a_`? That disables the test so it hasn't been running at all.

File: tests/test_manifests.py [hunk 6]
```diff
@@ -163,6 +169,20 @@ def a_test_slim_rpm_in_containers_manifest(stored_proc):
...
+        if pk in container_pks:
+            found_cpe = False
+            for ref in package["externalRefs"]:
+                if ref["referenceType"] == "cpe22Type":
+                    # The CPE matches the one returned by setup_products_and_rpm_in_containers()
+                    assert ref["referenceLocator"] == cpe
+                    found_cpe = True
+            # There was at least one CPE
+            assert found_cpe
+
```
Comment (juspence) [inline tests/test_manifests.py]: We're already testing this below, but I guess it doesn't hurt to add it here as well. You can also break early after you set `found_cpe = True`.

File: tests/test_model.py [hunk 7]
```diff
@@ -174,6 +175,32 @@ def test_component_cpes():
         assert unused_variant.cpe not in cpes
 
 
+@pytest.mark.django_db(databases=("default", "read_only"), transaction=True)
+def test_component_without_variant_cpes():
```
Comment (juspence) [inline tests/test_model.py:179]: Let's add a docstring comment:
```suggestion
@pytest.mark.django_db(databases=("default", "read_only"), transaction=True)
def test_component_without_variant_cpes():
    """Test the CPEs property on the Component model finds CPEs
        even when no Variants are linked directly to that Component"""
```

File: corgi/core/models.py (hunk 1 above)
Comment (jasinner) [reply corgi/core/models.py]: Good idea to also filter by type. Although I think filtering by ERRATA type is probably best since that is the best indicator that the build has been released to that Variant. Errata type relations shouldn't ever not have a linked SoftwareBuild foreign key, so leaving that part of the filter unchanged.

File: corgi/core/models.py (hunk 2 above)
Comment (jasinner) [reply corgi/core/models.py:1803]: ProductVariant models don't exist for brew_tag streams (like openshift-4.*) which we are targeting here. We would have make a bunch of other changes to the code to allow that linking. For example, setting up a many-to-many relationship between stream and variant to allow variants to be linked to multiple streams. Also changes to the save_product_taxonomy functions to prevent builds linked to ProductVariants matched via brew_tags from being also linked to ProductStreams. 

Using the Collector models here seems like a good compromise to me. Although if we really wanted to be strict about that we should probably allow ProductVariant models to have their productstreams, etc be null so that we can create ProductVariant and use those parentless ProductVariants in situations like this.

File: corgi/core/models.py (hunk 3 above)
Comment (jasinner) [reply corgi/core/models.py:1780]: I put the old comment back, it's still Variants.

File: corgi/web/templates/product_manifest.json (hunk 4 above)
Comment (jasinner) [reply corgi/web/templates/product_manifest.json]: The cpes variable is define only for ProductManifestFile in corgi/core/files.py. The point of this is to make sure that we're only specifying the component's CPEs where they the product's CPEs list. Components can ship to multiple products. We couldn't potentially report all CPEs for a components in the component_manifest.json, but that's beyond the scope of this PR in my opinion.

I moved the for, comment and if statement only a single line.

File: tests/test_manifests.py (hunk 5 above)
Comment (jasinner) [reply tests/test_manifests.py:136]: I don't know.

File: tests/test_manifests.py (hunk 6 above)
Comment (jasinner) [reply tests/test_manifests.py]: ok

File: tests/test_model.py (hunk 7 above)
Comment (jasinner) [reply tests/test_model.py:179]: ok

**Comment by jasinner** [comment]
> I'm not sure I understand the change, it seems incomplete. Right now we only report CPEs for root components, but I thought in the ticket we wanted to report CPEs for all components.
> 
> If that's correct, you can just add the `{% for cpe in component.cpes %}` logic to the externalRefs key for provided and upstream components as well. Right now this for-loop only exists for root components. This should be enough to start reporting CPEs for all components, changes to the Component model's "cpes" property shouldn't be needed.
> 
> Your changes to the "cpes" property will mostly work fine I think, but I'm not sure what edge-case you're trying to handle. I say "mostly" because self.software_build is only set for root components, and again I thought we were trying to report CPEs for all components.

You misunderstood the modivation for this change. The change to assist in determining which CPE a root component relates to when there are no linked ProductVariants for the component, for example, for openshift-4.x streams which use brew_tag relations to determine which builds make up the product.

File: corgi/core/models.py (hunk 2 above)
Comment (juspence) [reply corgi/core/models.py:1803]: Ack, using the Collector Variants is fine in this case. Thanks for explaining the edge-case we're handling with the brew_tag streams.

**Review by juspence — APPROVED** [review]
LGTM, thanks for explaining the problem with the brew_tag streams.

**Comment by jasinner** [comment]
CORGI-857 is fixed by the changes merged in PR#723, verified in: PR#732
